components/scheduler/executors.go

```diff
@@ -48,16 +68,70 @@ type (
 	}
 )
 
-func (e taskExecutor) executeScheduleTask(
+func (e taskExecutor) executeSchedulerWaitTask(
 	env hsm.Environment,
 	node *hsm.Node,
-	task ScheduleTask,
+	task SchedulerWaitTask,
 ) error {
 	if err := node.CheckRunning(); err != nil {
 		return err
 	}
-	// TODO(Tianyu): Perform scheduler logic before scheduling self again
-	return hsm.MachineTransition(node, func(scheduler Scheduler) (hsm.TransitionOutput, error) {
+	return hsm.MachineTransition(node, func(scheduler *Scheduler) (hsm.TransitionOutput, error) {
 		return TransitionSchedulerActivate.Apply(scheduler, EventSchedulerActivate{})
 	})
 }
+
+func (e taskExecutor) executeSchedulerRunTask(
+	ctx context.Context,
+	env hsm.Environment,
+	ref hsm.Ref,
+	task SchedulerActivateTask,
+) error {
+	return env.Access(ctx, ref, hsm.AccessWrite, func(node *hsm.Node) error {
+		if err := node.CheckRunning(); err != nil {
+			return err
+		}
+		s, err := hsm.MachineData[*Scheduler](node)
+		if err != nil {
+			return err
+		}
+
+		s.populateTransientFieldsIfAbsent(e.options.Logger, e.options.MetricsHandler, e.options.SpecBuilder, e.options.FrontendClient, e.options.HistoryClient)
+
+		if s.Args.State.LastProcessedTime == nil {
+			// log these as json since it's more readable than the Go representation
+			specJson, _ := protojson.Marshal(s.Args.Schedule.Spec)
+			policiesJson, _ := protojson.Marshal(s.Args.Schedule.Policies)
+			s.logger.Info("Starting schedule", tag.NewStringTag("spec", string(specJson)), tag.NewStringTag("policies", string(policiesJson)))
+			s.Args.State.LastProcessedTime = timestamppb.Now()
+			s.Args.State.ConflictToken = scheduler.InitialConflictToken
+			s.Args.Info.CreateTime = s.Args.State.LastProcessedTime
+		}
+
+		t1 := timestamp.TimeValue(s.Args.State.LastProcessedTime)
+		t2 := time.Now()
+		if t2.Before(t1) {
+			s.logger.Warn("Time went backwards", tag.NewStringTag("time", t1.String()), tag.NewStringTag("time", t2.String()))
+			t2 = t1
+		}
+		nextWakeup, lastAction := s.processTimeRange(
+			t1, t2,
+			// resolve this to the schedule's policy as late as possible
+			enumspb.SCHEDULE_OVERLAP_POLICY_UNSPECIFIED,
+			false,
+			nil,
+		)
+		s.Args.State.LastProcessedTime = timestamppb.New(lastAction)
+		// process backfills if we have any too
+		s.processBackfills()
+		// try starting workflows in the buffer
+		//nolint:revive
+		for s.processBuffer() {
+		}
```

Comment:
All of these calls should be done outside of the `Access` block to avoid locking the workflow mutable state while making network calls.

See the callback implementation for reference.